src/hotspot/share/gc/shenandoah/shenandoahOldGC.cpp

```diff
@@ -111,6 +111,12 @@ bool ShenandoahOldGC::collect(GCCause::Cause cause) {
   // Complete marking under STW
   vmop_entry_final_mark();
 
+  if (_generation->is_concurrent_mark_in_progress()) {
+    assert(heap->cancelled_gc(), "Safepoint operation observed gc cancellation");
+    // GC may have been cancelled before final mark, but after the preceding cancellation check.
+    return false;
+  }
+
```

Comment:
This looks like it might be a bug fix, vs the rest of this PR, which is mostly isolation refactoring.  Should the bug fix be separated out into a different PR?

Reply:
Yes, it is a bug fix. I'll pull it out into a separate PR.

Reply:
Separate PR for this here: https://github.com/openjdk/shenandoah/pull/448

Reply:
Thanks for pulling this out separately. I do think this is subtle and we should (in the fullness of time) make this idiom a bit cleaner and less fragile and uniform in terms of the representation of the state machine that can be used by both the concurrent and the degenerate resumptions.